### PR TITLE
Port to use crossbeam instead of mpsc_select

### LINKED ISSUE #45: mpsc is being deprecated/removed
See the discussion here: https://github.com/rust-lang/rust/issues/27800#issuecomment-421760863

We should look into whether or not we can use [crossbeam channels](https://github.com/crossbeam-rs/crossbeam-channel) instead, since they seem to be the replacement of choice. Other options may be viable as well. That'll also help us actually have something useful that works on stable.

### PR DESCRIPTION
**Pull request by Munksgaard**
Fixes #45 

Todo:
 - [x] hselect
 - [x] iselect
 - [x] ChanSelect
 - [x] offer!
 - [x] try_offer!
 - [x] chan_select!

### CODE REVIEW

File: examples/hselect.rs [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+
+use session_types::*;
+
+fn main() {
+    let (tcs, rcs) = session_channel();
+    let (tcu, rcu) = session_channel();
+
+    let receivers = vec!(rcs, rcu);
+
+    let () = tcs.send("Hello, World from TCS!".to_string()).close();
```
Comment (laumann) [inline examples/hselect.rs]: This pattern looks odd to me, `let () = …`. I thought the idiomatic way to discard results is `let _ = …`. In this case, I don't think calls to `.close()` have `#[must_use]`

File: src/lib.rs [hunk 2]
```diff
@@ -134,6 +131,24 @@ pub struct Rec<P>(PhantomData<P>);
 /// out of.
 pub struct Var<N>(PhantomData<N>);
 
+trait HasReceiver {
```
Comment (laumann) [inline src/lib.rs]: Why is this necessary? To allow pulling the `Receiver<_>` out of different channels types?

EDIT Doesn't appear to be used.

File: src/lib.rs [hunk 3]
```diff
@@ -402,36 +416,22 @@ pub fn hselect<E, P, A>(
...
-            }
-        }
+        let id = sel.ready();
 
-        let id = sel.wait();
-
-        for handle in handles.iter_mut() {
-            // Clean up
-            unsafe {
-                handle.remove();
```
Comment (laumann) [inline src/lib.rs]: It's gonna be nice to get rid of this :+1: It was always a bit yucky :stuck_out_tongue:

File: src/lib.rs [hunk 4]
```diff
@@ -447,89 +447,68 @@ pub fn iselect<E, P, A>(chans: &Vec<Chan<E, Recv<A, P>>>) -> usize {
...
-/// of the selected channel. This is also the implementation used by
-/// the `chan_select!` macro.
-#[cfg(feature = "chan_select")]
-impl<'c> ChanSelect<'c, usize> {
-    pub fn add_recv<E, P, A: marker::Send>(&mut self, c: &'c Chan<E, Recv<A, P>>) {
-        let index = self.chans.len();
-        self.add_recv_ret(c, index);
-    }
+// /// Default use of ChanSelect works with usize and returns the index
+// /// of the selected channel. This is also the implementation used by
```
Comment (laumann) [inline src/lib.rs]: Just remove commented code.

File: src/lib.rs [hunk 5]
```diff
@@ -575,7 +554,7 @@ mod private {
 /// we can use the `offer!` macro as follows:
 ///
 /// ```rust
-/// #[macro_use] extern crate session_types;
```
Comment (laumann) [inline src/lib.rs]: I'm curious: Does this break compatibility with older versions of rustc?

File: tests/chan_select.rs [hunk 6]
```diff
@@ -62,13 +82,8 @@ fn chan_select_simple() {
     recv_assert_eq_close!("Hello, World!".to_owned(), rcs.recv());
 }
 
-#[cfg(feature = "chan_select")]
 #[test]
 fn chan_select_add_ret() {
```
Comment (laumann) [inline tests/chan_select.rs]: I'm kinda thinking this test could just be removed - the whole idea with the `_ret` stuff was to allow users to select the types returned from `sel.wait()` (in this test the `ChanToRead`). But that whole thing is gone now, so I'm not sure this test adds anything.

**Comment by Munksgaard** [comment]
> I've just published `crossbeam-channel` version 0.3.3 with the new features so you can add it to `Cargo.toml`.

That's great news. Thank you @stjepang :+1:

**Comment by Munksgaard** [comment]
@laumann What do you think, should we merge it? I think I've addressed all of your comments. 

I'm thinking that this probably necessitates a bump to 0.3.0 for us. While the outward API hasn't changed, the crate is suddenly fully usable on stable and you don't need to use the `chan_select` feature any more.

File: .travis.yml [hunk 7]
```diff
@@ -16,7 +16,7 @@ script:
 
 env:
   global:
-    - TRAVIS_CARGO_NIGHTLY_FEATURE=chan_select
+    - TRAVIS_CARGO_NIGHTLY_FEATURE=""
```
Comment (laumann) [inline .travis.yml]: Can this be removed?

File: examples/hselect.rs [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+
+    let _ = tcs.send("Hello, World from TCS!".to_string()).close();
+
+    let (ready, mut rest) = hselect(receivers);
+
+    let (to_close, s) = ready.recv();
+    println!("Got a response: \"{}\"", s);
+    to_close.close();
+
+    let _ = tcu.send("Hello, World from TCU!".to_string()).close();
```
Comment (laumann) [inline examples/hselect.rs]: I think all the call to `close()` do not need the `let _ = …`, ie

```
tcu.send("Hello, World from TCU!".to_string()).close();
```
should suffice.

(This applies in a few places.)

File: tests/chan_select.rs [hunk 9]
```diff
@@ -19,7 +14,32 @@ macro_rules! recv_assert_eq_close(
...
 
-#[cfg(feature = "chan_select")]
+#[test]
+fn chan_select_hselect() {
+    let (tcs, rcs) = session_channel();
+    let (tcu, rcu) = session_channel();
+
+    let receivers = vec!(rcs, rcu);
+
+    let () = tcs.send(1u64).close();
```
Comment (laumann) [inline tests/chan_select.rs]: There's `let () = …` in a few places that I think could just be removed.

File: tests/tests.rs [hunk 10]
```diff
@@ -1,22 +1,17 @@
...
-#[cfg(feature = "chan_select")]
 fn run_mode(mode: &'static str) {
     let mut config = compiletest::Config::default();
-    let cfg_mode = mode.parse().ok().expect("Invalid mode");
 
-    config.mode = cfg_mode;
+    config.mode = mode.parse().expect("Invalid mode");
     config.src_base = PathBuf::from(format!("tests/{}", mode));
-    config.target_rustcflags = Some("-L target/debug".to_string());
+    config.link_deps();
```
Comment (laumann) [inline tests/tests.rs]: Using `link_deps()` has a tendency to break on Windows, because of spaces in paths, so I'd be careful when using it :-)

File: .travis.yml (hunk 7 above)
Comment (Munksgaard) [reply .travis.yml]: No, if we don't define `TRAVIS_CARGO_NIGHTLY_FEATURE` it is automatically set to `unstable`, which is not a feature in our library. See the docs here: https://github.com/huonw/travis-cargo

File: tests/tests.rs (hunk 10 above)
Comment (Munksgaard) [reply tests/tests.rs]: The tests stop working on my machine if I replace `config.link_deps()` with `config.target_rustcflags = Some("-L target/debug".to_string());`. Any idea why?
